### PR TITLE
8339803: Acknowledge case insensitive unambiguous keywords in tzdata files

### PR DESCRIPTION
**Pull request by naotoj**
This is a follow on fix to [JDK-8339644](https://bugs.openjdk.org/browse/JDK-8339644). It turned out that TZ data files allow abbreviation of keywords, such as "ZO" for "Zone." Same fix, i.e, replacing `startsWith()` with `regionMatches()` was applied.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8339803](https://bugs.openjdk.org/browse/JDK-8339803): Acknowledge case insensitive unambiguous keywords in tzdata files (**Bug** - P4)

### Reviewers
 * [Justin Lu](https://openjdk.org/census#jlu) (@justin-curtis-lu - Committer)
 * [Sean Coffey](https://openjdk.org/census#coffeys) (@coffeys - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/20940/head:pull/20940` \
`$ git checkout pull/20940`

Update a local copy of the PR: \
`$ git checkout pull/20940` \
`$ git pull https://git.openjdk.org/jdk.git pull/20940/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 20940`

View PR using the GUI difftool: \
`$ git pr show -t 20940`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/20940.diff">https://git.openjdk.org/jdk/pull/20940.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/20940#issuecomment-2341890517)

### CODE REVIEW

File: make/jdk/src/classes/build/tools/cldrconverter/CLDRConverter.java [hunk 1]
```diff
@@ -1372,6 +1372,7 @@ private static Map<Locale, String> coverageLevelsMap() throws Exception {
     private static void generateTZDBShortNamesMap() throws IOException {
         Files.walk(Path.of(tzDataDir), 1, FileVisitOption.FOLLOW_LINKS)
             .filter(p -> p.toFile().isFile())
+            .filter(p -> p.getFileName().toString().matches("africa|antarctica|asia|australasia|backward|backzone|etcetera|europe|factory|northamerica|southamerica"))
```
Comment (justin-curtis-lu) [inline make/jdk/src/classes/build/tools/cldrconverter/CLDRConverter.java:1375]: I am just curious, why is this needed?

Comment (naotoj) [reply make/jdk/src/classes/build/tools/cldrconverter/CLDRConverter.java]: It would otherwise scan all files under the directory, and fails with `iso3166.tab` or `zone.tab`, which include "LI" as the country code at the beginning of the line, that interfere with "Link" match.
Having said that, the default list was different from the one in `TzdbZoneRulesProvider`. I removed extra file names.

Comment (justin-curtis-lu) [reply make/jdk/src/classes/build/tools/cldrconverter/CLDRConverter.java]: Right, the match in general is much more lenient now, so need to restrict the files searched for false positives. Thanks for the explanation, that makes sense.

**Review by justin-curtis-lu — APPROVED** [review]
Looks good to me

**Comment by naotoj** [comment]
Thanks, @justin-curtis-lu. @coffeys, would you please review this PR, which can also apply to the tzupdater tool?

**Comment by coffeys** [comment]
LGTM

`sun/util/calendar/zi/TestZoneInfo310.java` exercises the tzdb compiler. I wonder if it's worth adding some dummy rules which test the new leniency added to the compiler ? (perhaps even just to the jdk11_backward file included in the test dir)

**Comment by naotoj** [comment]
Thanks, @coffeys 

> `sun/util/calendar/zi/TestZoneInfo310.java` exercises the tzdb compiler. I wonder if it's worth adding some dummy rules which test the new leniency added to the compiler ? (perhaps even just to the jdk11_backward file included in the test dir)

IIUC the purpose of that test is to check the integrity of `tzdb.dat` which is created at the JDK build time, with the pre-310 era javazic. So adding extra tz files at the test time would break (because `tzdb.dat` do not include that lenient test tz file). Testing of JDK build tool is hard IMO.

/integrate

**Review by coffeys — APPROVED** [review]
ah yes, it's the other way around then (java zi compiler creating legacy data to match JSR 310 data) - ok, impossible to test then..  outside of setting up a build test env.

maybe you can verify with a few simple edits to a tzdata file and building.

**Comment by naotoj** [comment]
> maybe you can verify with a few simple edits to a tzdata file and building

Yes, that is exactly what I did for this testing.

/integrate

File: make/jdk/src/classes/build/tools/tzdb/TzdbZoneRulesProvider.java [hunk 2]
```diff
@@ -304,7 +305,7 @@ private void parse(String[] tokens, int off) {
             month = parseMonth(tokens[off++]);
             if (off < tokens.length) {
                 String dayRule = tokens[off++];
-                if (dayRule.startsWith("last")) {
+                if (dayRule.regionMatches(true, 0, "last", 0, 4)) {
```
Comment (jodastephen) [inline make/jdk/src/classes/build/tools/tzdb/TzdbZoneRulesProvider.java:308]: This isn't correct, as per the mailing list:

> > can "last" contain uppercase letters, or
> > does it have to be exactly "last"?
> 
> In that case the name is "lastSunday", and it can be abbreviated
> "lastSu" or "Lastsu" or "LASTSUNDA" or whatever.

Comment (naotoj) [reply make/jdk/src/classes/build/tools/tzdb/TzdbZoneRulesProvider.java:308]: Why is it not? IIUC, abbreviation should be unambiguous, so "last" part is always case-insensitive 4 character length.

Comment (jodastephen) [reply make/jdk/src/classes/build/tools/tzdb/TzdbZoneRulesProvider.java:308]: Sorry, I assumed this regionMatches was the same as the others. "last" is always "last" and never abbreviated.
